### PR TITLE
feat(listbox): add grouping to listbox

### PR DESCRIPTION
**Pull request by 321gillian**
Add grouping to listbox

GDS-2287

### CODE REVIEW

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 1]
```diff
@@ -38,60 +70,32 @@ function setActiveOption(
   }
 }
 
-function getFirstOption(list: HTMLGuxListboxElement): ListboxOptionElement {
-  let firstOption = getListOptions(list)[0];
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:42]: `return getAvailableListOptions(list)[0];`

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 2]
```diff
@@ -38,60 +70,32 @@ function setActiveOption(
...
-    }
-
-    return previousOption;
-  }
-
-  return getFirstOption(list);
-}
-
-function getLastOption(list: HTMLGuxListboxElement): ListboxOptionElement {
-  const options = getListOptions(list);
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:86]:
```
const availableOptions =  getAvailableListOptions(list);
return availableOptions[availableOptions.length - 1];
```

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 3]
```diff
@@ -38,60 +70,32 @@ function setActiveOption(
...
-  let firstOption = getListOptions(list)[0];
-
-  while (firstOption && (firstOption.disabled || firstOption.filtered)) {
-    firstOption = firstOption.nextElementSibling as ListboxOptionElement;
-  }
-
-  return firstOption;
-}
-
-function getNextOption(list: HTMLGuxListboxElement): ListboxOptionElement {
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:51]:
```
  if (hasActiveOption(list)) {
    const availableListOptions = getAvailableListOptions(list);
    const activeOption = getActiveOption(list);
    const activeOptionIndex = availableListOptions.indexOf(activeOption);

   return availableOptionsList[activeOptionIndex + 1]
  }

  return getFirstOption(list);
```

I am missing a check if you are on the last option in the above that you should add

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 4]
```diff
@@ -38,60 +70,32 @@ function setActiveOption(
...
-      nextOption = nextOption.nextElementSibling as ListboxOptionElement;
-    }
-
-    return nextOption;
-  }
-
-  return getFirstOption(list);
-}
-
-function getPreviousOption(list: HTMLGuxListboxElement): ListboxOptionElement {
```
Comment (daragh-king-genesys) [inline packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:66]:
```
  if (hasActiveOption(list)) {
    const availableListOptions = getAvailableListOptions(list);
    const activeOption = getActiveOption(list);
    const activeOptionIndex = availableListOptions.indexOf(activeOption);

   return availableOptionsList[activeOptionIndex - 1]
  }

  return getFirstOption(list);
```
I am missing a check if you are on the first option in the above that you should add

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 5]
```diff
@@ -1,13 +1,45 @@
...
+export function isOptionGroup(item: Element): boolean {
+  return item.tagName === 'GUX-OPTION-GROUP-BETA';
+}
+
+export function isOption(item: Element): boolean {
+  const optionTypes = ['GUX-OPTION', 'GUX-OPTION-ICON'];
+  return optionTypes.includes(item.tagName);
+}
+
+export function hasOptionGroups(list: HTMLGuxListboxElement): boolean {
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts]: removed

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts [hunk 6]
```diff
@@ -38,60 +70,32 @@ function setActiveOption(
...
-}
-
-function getLastOption(list: HTMLGuxListboxElement): ListboxOptionElement {
-  const options = getListOptions(list);
-
-  let lastOption = options[options.length - 1];
-
-  while (lastOption && (lastOption.disabled || lastOption.filtered)) {
-    lastOption = lastOption.previousElementSibling as ListboxOptionElement;
+function getListOptionByLocation(
```
[the remark this answers is not in the document]
Comment (321gillian) [reply packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts]: done

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts (hunk 4 above)
Comment (321gillian) [reply packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:66]: Done

File: packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts (hunk 3 above)
Comment (321gillian) [reply packages/genesys-spark-components/src/components/stable/gux-listbox/gux-listbox.service.ts:51]: Done

**Review by daragh-king-genesys — APPROVED** [review]

**Pull request merged**